### PR TITLE
JdbcTemplate CRUD.

### PR DESCRIPTION
**Pull request by YevhenDov**

### CODE REVIEW

File: SpingJdbc/pom.xml [hunk 1]
```diff
@@ -0,0 +1,96 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>SpringJdbc</groupId>
+    <artifactId>SpingJdbc</artifactId>
+    <packaging>jar</packaging>
```
Comment (Anderwerty) [inline SpingJdbc/pom.xml]: Why jar?

File: SpingJdbc/pom.xml [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-beans</artifactId>
+            <version>5.1.5.RELEASE</version>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-aop</artifactId>
+            <version>5.1.4.RELEASE</version>
+        </dependency>
+
```
Comment (Anderwerty) [inline SpingJdbc/pom.xml]: Remove redundant blank lines.

File: SpingJdbc/pom.xml [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+        <dependency>
+            <groupId>mysql</groupId>
+            <artifactId>mysql-connector-java</artifactId>
+            <version>8.0.15</version>
+        </dependency>
+
```
Comment (Anderwerty) [inline SpingJdbc/pom.xml]: The same. Please use one style for the whole pom.xml file.

File: SpingJdbc/src/main/java/com/annotation/InjectRandomInt.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package com.annotation;
```
Comment (Anderwerty) [inline SpingJdbc/src/main/java/com/annotation/InjectRandomInt.java]: It looks like your root package is com.
Сom is a top-level domain. Please add a root package
Example: com.google, com.private24

File: SpingJdbc/src/main/java/com/bpp/MyBeanPostProcessor.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    public Object postProcessBeforeInitialization(Object bean, String beanName) throws BeansException {
+
+        Field[] fields = bean.getClass().getFields();
+        for (Field field : fields) {
+            InjectRandomInt annotation = field.getAnnotation(InjectRandomInt.class);
+            if (annotation != null) {
+                int min = annotation.min();
+                int max = annotation.max();
+                Random random = new Random();
+                int i = min + random.nextInt(max - min);
```
Comment (Anderwerty) [inline SpingJdbc/src/main/java/com/bpp/MyBeanPostProcessor.java]: `i `is not informative name for the variable.

File: SpingJdbc/src/main/java/com/dao/daoimpl/UserDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+import com.entity.User;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.jdbc.core.BeanPropertyRowMapper;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.stereotype.Component;
+
+import javax.sql.DataSource;
+import java.util.Optional;
+
+@Component("userDaoImpl")
```
Comment (Anderwerty) [inline SpingJdbc/src/main/java/com/dao/daoimpl/UserDaoImpl.java]: It is better to use specific annotation
@Repository
http://qaru.site/questions/1316/whats-the-difference-between-component-repository-service-annotations-in-spring

File: SpingJdbc/src/main/java/com/dao/daoimpl/UserDaoImpl.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+    @Override
+    public void createUser(User user) {
+        jdbcTemplate.update(CREATE, new Object[]{user.getName(), user.getAge()});
+    }
+
+    @Override
+    public Optional<User> getUserById(int id) {
+        User user = jdbcTemplate.queryForObject(READ, new Object[]{id}, new BeanPropertyRowMapper<>(User.class));
+        Optional<User> optionalUser = Optional.of(user);
```
Comment (Anderwerty) [inline SpingJdbc/src/main/java/com/dao/daoimpl/UserDaoImpl.java]: If in line 34 you get null, in line 35 you will get an exception.

File: SpingJdbc/src/main/java/com/dao/daoimpl/UserDaoImpl.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+public class UserDaoImpl implements UserDao {
+    private static final String CREATE = "INSERT INTO users (name, age) VALUES (?, ?)";
+    private static final String READ = "SELECT * FROM users WHERE id = ?";
+    private static final String UPDATE = "UPDATE users SET name = ?, age = ? WHERE id = ?";
+    private static final String DELETE = "DELETE FROM Users WHERE id = ?";
+
+    private JdbcTemplate jdbcTemplate;
+
+    @Autowired
+    public void setJdbcTemplate(DataSource dataSource) {
```
Comment (Anderwerty) [inline SpingJdbc/src/main/java/com/dao/daoimpl/UserDaoImpl.java]: Why not use the constructor for autowired?

File: SpingJdbc/src/main/java/com/entity/User.java [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.GeneratedValue;
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+
+@Entity
+@Data
+@Component("user")
```
Comment (Anderwerty) [inline SpingJdbc/src/main/java/com/entity/User.java]: Why did you add this annotation here?

File: SpingJdbc/src/test/java/com/dao/daoimpl/UserDaoImplTest.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+        userActual.setAge(15);
+        userActual.setName("John");
+        userDao.createUser(userActual);
+
+        userDao.deleteUser(2);
+        userActual = userDao.getUserById(2).get();
+
+        assertEquals(userExpected, userActual);
+    }
+}
```
Comment (Anderwerty) [inline SpingJdbc/src/test/java/com/dao/daoimpl/UserDaoImplTest.java]: Add a blank line at the end of the file.

File: SpingJdbc/src/test/java/com/dao/daoimpl/UserDaoImplTest.java [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+import com.entity.User;
+import org.junit.Before;
+import org.junit.Test;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.support.ClassPathXmlApplicationContext;
+
+import static org.junit.Assert.assertEquals;
+
+public class UserDaoImplTest {
```
Comment (Anderwerty) [inline SpingJdbc/src/test/java/com/dao/daoimpl/UserDaoImplTest.java]: Did you run tests on production database?

File: SpingJdbc/src/main/java/com/entity/User.java (hunk 9 above)
Comment (YevhenDov) [reply SpingJdbc/src/main/java/com/entity/User.java]: it for tests

File: SpingJdbc/src/test/java/com/dao/daoimpl/UserDaoImplTest.java (hunk 11 above)
Comment (YevhenDov) [reply SpingJdbc/src/test/java/com/dao/daoimpl/UserDaoImplTest.java]: yes. I can upload screens

File: SpingJdbc/src/main/java/com/entity/User.java (hunk 9 above)
Comment (Anderwerty) [reply SpingJdbc/src/main/java/com/entity/User.java]: You should not put annotation into code just for tests!

File: SpingJdbc/src/test/java/com/dao/daoimpl/UserDaoImplTest.java (hunk 11 above)
Comment (Anderwerty) [reply SpingJdbc/src/test/java/com/dao/daoimpl/UserDaoImplTest.java]: You should not run tests on the production database. In a such way you could delete necessary information.

File: SpingJdbc/src/main/java/com/company/controller/CrudController.java [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+import org.springframework.stereotype.Controller;
+import org.springframework.ui.Model;
+import org.springframework.web.bind.annotation.ModelAttribute;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RequestParam;
+
+@Controller
+public class CrudController {
+    private UserServiceImpl userService;
```
Comment (Anderwerty) [inline SpingJdbc/src/main/java/com/company/controller/CrudController.java]: How did you inject service into the controller?
And why reference type is not interface?

File: SpingJdbc/src/main/java/com/company/servce/impl/UserServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+
+import javax.annotation.PostConstruct;
+import java.util.Optional;
+
+@Log4j
+@Service
+public class UserServiceImpl implements UserService {
+    private static final String PATH_TO_FILE = "META-INF/spring/context.xml";
+
+    private UserDaoImpl userDao;
```
Comment (Anderwerty) [inline SpingJdbc/src/main/java/com/company/servce/impl/UserServiceImpl.java]: Why type of reference is not interface but implementation?

File: SpingJdbc/src/main/java/com/company/servce/impl/UserServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+
+    @PostConstruct
+    public void init() {
+        ApplicationContext context = new ClassPathXmlApplicationContext(PATH_TO_FILE);
+        userDao = (UserDaoImpl) context.getBean("userDaoImpl");
```
Comment (Anderwerty) [inline SpingJdbc/src/main/java/com/company/servce/impl/UserServiceImpl.java]: Please use the overload method getBean.

File: SpingJdbc/src/main/java/com/company/servce/UserService.java [hunk 15]
```diff
@@ -0,0 +1,15 @@
...
+import com.company.entity.User;
+
+import java.util.Optional;
+
+public interface UserService {
+    void createUser(User user);
+
+    Optional<User> getUserById(int id);
+
+    void updateUser(int id, User user);
```
Comment (Anderwerty) [inline SpingJdbc/src/main/java/com/company/servce/UserService.java]: Why do you need parameter id? Why not put it into user

Comment (YevhenDov) [reply SpingJdbc/src/main/java/com/company/servce/UserService.java]: 'id' for found by id, 'user' for update

Comment (Anderwerty) [reply SpingJdbc/src/main/java/com/company/servce/UserService.java]: if you update user you should not change id.

File: SpingJdbc/src/main/java/com/company/dao/daoimpl/UserDaoImpl.java [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    @Override
+    public void createUser(User user) {
+        jdbcTemplate.update(CREATE, new Object[]{user.getName(), user.getAge()});
+    }
+
+    @Override
+    public Optional<User> getUserById(int id) {
+        Optional<User> optionalUser = Optional.of(jdbcTemplate.queryForObject(READ, new Object[]{id}, new BeanPropertyRowMapper<>(User.class)));
```
Comment (Anderwerty) [inline SpingJdbc/src/main/java/com/company/dao/daoimpl/UserDaoImpl.java]: queryForObject() throws EmptyResultDataAccessException when record not found
It would help you to fix it
https://www.mkyong.com/spring/queryforobject-throws-emptyresultdataaccessexception-when-record-not-found/
Please fix

File: SpingJdbc/src/main/java/com/company/servce/UserService.java (hunk 15 above)
Comment (YevhenDov) [reply SpingJdbc/src/main/java/com/company/servce/UserService.java]: I don't edit. It in the query where the parameter 'where'

File: SpingJdbc/src/main/java/com/company/dao/daoimpl/UserDaoImpl.java [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+
+    @Override
+    public void createUser(User user) {
+        jdbcTemplate.update(CREATE, new Object[]{user.getName(), user.getAge()});
+    }
+
+    @Override
+    public Optional<User> getUserById(int id) {
+        try {
+            Optional<User> optionalUser = Optional.of(jdbcTemplate.queryForObject(READ, new Object[]{id}, new BeanPropertyRowMapper<>(User.class)));
```
Comment (Anderwerty) [inline SpingJdbc/src/main/java/com/company/dao/daoimpl/UserDaoImpl.java]: Just
return Optional.of(jdbcTemplate.queryForObject(READ, new Object[]{id}, new BeanPropertyRowMapper<>(User.class)));

File: SpingJdbc/src/main/java/com/company/dao/daoimpl/UserDaoImpl.java [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+            return optionalUser;
+        } catch (EmptyResultDataAccessException e){
+            return null;
```
Comment (Anderwerty) [inline SpingJdbc/src/main/java/com/company/dao/daoimpl/UserDaoImpl.java]: return not null  but 
Optional.empty()
